### PR TITLE
Experiment with static past key/value buffer

### PR DESCRIPTION
**Pull request by fxmarty**
This PR is just to see if this could reside in transformers or not.

## Motivation

We suspect [the concatenations of the key/value buffer](https://github.com/huggingface/transformers/blob/ba695c1efd55091e394eb59c90fb33ac3f9f0d41/src/transformers/models/gpt2/modeling_gpt2.py#L321-L322) at each generation step to be expensive. The idea is to modify the buffer in place after a unique allocation.

FasterTransformer [does preallocate the kv cache](https://github.com/NVIDIA/FasterTransformer/blob/c6e8f60ec40da218804a60e6aa986903e7fa8594/src/fastertransformer/models/decoding/Decoding.cc#L83C9-L84) (among others). Preallocating may also help `torch.compile` according to @Chillee, although I don't quite get why yet (there are still dynamic shapes in the model itself, so why care about model I/O static shapes?).

For reference: https://huggingface.slack.com/archives/C055NT312LW/p1683038064467109 & https://huggingface.slack.com/archives/C055NT312LW/p1685570357532069

## Current (ugly) workflow

Very temporary

```python
cache_size = max_new_tokens

past_key_values = tuple([
    tuple([torch.empty(
        batch_size,
        model.config.n_head,
        cache_size,
        model.config.n_embd // model.config.n_head,  # head dimension
        dtype=dtype,
        device=device
    ) for _ in range(2)])

model.enable_static_kv_cache()
res = model.generate(
    **inputs,
    num_beams=1,
    min_new_tokens=max_new_tokens,
    max_new_tokens=max_new_tokens,
    past_key_values=past_key_values
)
```

## Results

Some optimizations may still be missing - right now in small model / small batch size setting this is not interesting.

Results are with PyTorch 2.0.1 eager.

Script: https://github.com/fxmarty/transformers-preallocate-kv-cache/blob/main/run.py
Raw results: https://docs.google.com/spreadsheets/d/15P1o9vDcXOSeLAwLUWiqXQatHxtQYBa_xcyLZanOjQQ/edit?usp=sharing

image
image
image
image
image
image

## Misc

The current implementation of `valid_past_index` being `Optional[int]` is quite debatable and may hamper readability.

I'm also not sure whether adding methods in `PreTrainedModel` that are specific to decoders is OK?

Missing test right now is to generate a sequence length shorter than the KV buffer.

Some todos:
- [ ] `past_key_values` buffer should be initialized in the background, instead of requiring the user to initialize it himself and requiring to pass `generate(**inputs, past_key_values=past_key_values)` (as currently).
- [ ] Current implementation is likely to break with `accelerate` with naive pipeline parallelism, as the buffer is currently initialized on a single device
- [ ] Preallocated kv cache still does not help with small models / batch size.
- [ ] Support an iterative buffer, e.g. that auto-extends each 512 tokens, instead of initializing a buffer of size `max_new_tokens`. This may help reducing memory usage (and speed? probably not).
- [ ] Implement tests
- [ ] Should this be in optimum or transformers?
- [ ] Support all (is it possible?) decoding strategies, instead of currently only `greedy_search`
- [ ] Have it work with cross-attention
- [ ] Preallocate `attention_mask` as well
- [ ] Test on CPU as well

### CODE REVIEW

**Review by younesbelkada — COMMENTED** [review]
For me this is a great addition and I would advocate to support it directly in transformers! 
Regarding the API maybe the cache pre-allocation can be done directly on `generate` (after wrapping this in a method on GPT2PreTrainedModel maybe `def _preallocate_static_kv_cache`?). If the concept gets validated, happy to help moving forward with fixing the current CI tests and adding new tests for that as well
WDYT @gante @patrickvonplaten @sgugger @amyeroberts

File: src/transformers/models/gpt2/modeling_gpt2.py [hunk 1]
```diff
@@ -316,15 +317,29 @@ def forward(
...
+                key = torch.cat((past_key, key), dim=-2)
+                value = torch.cat((past_value, value), dim=-2)
+            if use_cache is True:
+                present = (key, value)
+        elif valid_past_index > 0:
+            upper = valid_past_index + 1
+            layer_past[0][:, :, valid_past_index:upper] = key  # Slice to keep dimension info
+            layer_past[1][:, :, valid_past_index:upper] = value
+
+            # TODO (felix) Would removing these two lines help?
```
Comment (amyeroberts) [inline src/transformers/models/gpt2/modeling_gpt2.py:332]: Help with what? :D

**Review by gante — COMMENTED** [review]
My 2 cents here:
1 - It requires extensive model code changes 😱 the speedups are nice in some settings, but IMO it will require a lot of work to update relevant architectures and it will likely expose us to unforeseen bugs. Since there are other PT2.0-related gains to be explored, I'd push these changes to a later stage of our optimizations 🤔 
2 - You wrote in the PR header `there are still dynamic shapes in the model itself, so why care about model I/O static shapes?`. Like in XLA, large speedup gains with `.compile()` are obtained from the "compile once run many times" paradigm. An expensive recompilation is triggered by different input shapes (regardless of the output and internal shapes), so static input shapes = 1 compilation = fast. Some compilations in the internal kernels may still happen due to the dynamic shapes present in the model, but the bulk of the work is in compiling the whole `forward` function (or better yet, the whole `generate` function). We got large speedups on the TF side with no model changes (e.g. keeping these internal concat operations), simply by passing static shapes to `generate` (and `forward`, by extension) 🙌

**Comment by patrickvonplaten** [comment]
In order to efficiently work with the PyTorch team here and to figure out exactly what needs to be done for a super fast generate method, I'd suggest to open a new benchmark repo that includes this PR for say three important LLM models:
- GPT2
- Llama
- starcoder

The repo "maybe just called `transformers-generate-benchmark` should have:
- copied the modeling files of GPT2, Llama, Starcoder
- Use the SDPA attention for all models (just like `betterTransformers` does - you could copy it from here: https://github.com/huggingface/optimum/blob/main/optimum/bettertransformer/models/attention.py)
- A super small generate function that we write ourselves (just greedy without any logit processor)

Then in this repo we run all the benchmarking and also make it easy for the PyTorch team to reproduce the benchmarking numbers.

**Comment by fxmarty** [comment]
Sounds good!

**Pull request closed without merging**